### PR TITLE
[CPU] Add implementation for quantized softmax

### PR DESCRIPTION
**Pull request by EpureRares**
Summary:
Add support for quantized softmax operator.

Test Plan:
Add unit test in OperatorTest

Co-author: iancu.mindru@nxp.com

### CODE REVIEW

File: include/glow/Quantization/Base/Base.h [hunk 1]
```diff
@@ -118,6 +118,92 @@ struct NodeQuantizationInfo {
...
+
+  /// \returns the number of fractional bits.
+  unsigned getFracBits() const { return fracBits_; }
+
+private:
+  // \p number to convert
+  // \returns the minimum number of bits for integer part of the
+  // fixed point representation
+  uint32_t minBitsIntegerPart(float number) {
+
```
Comment (mciprian13) [inline include/glow/Quantization/Base/Base.h:175]: Assert that ```number``` is positive.

File: include/glow/Quantization/Base/Base.h [hunk 2]
```diff
@@ -118,6 +118,92 @@ struct NodeQuantizationInfo {
...
+    uint32_t integerPart = 0;
+
+    while (aux / 2 != 0) {
+      integerPart += 1;
+      aux /= 2;
+    }
+
+    assert(integerPart >= 0 && integerPart < 32 &&
+           "Overflow caused by input number\n");
+    return integerPart + 1;
```
Comment (mciprian13) [inline include/glow/Quantization/Base/Base.h:186]: Why return integerPart+1?

File: include/glow/Quantization/Base/Base.h [hunk 3]
```diff
@@ -118,6 +118,92 @@ struct NodeQuantizationInfo {
...
+
+    uint32_t aux = (uint32_t)number;
+    uint32_t integerPart = 0;
+
+    while (aux / 2 != 0) {
+      integerPart += 1;
+      aux /= 2;
+    }
+
+    assert(integerPart >= 0 && integerPart < 32 &&
```
Comment (mciprian13) [inline include/glow/Quantization/Base/Base.h:184]: Can we have <= 32?

File: include/glow/Quantization/Base/Base.h [hunk 4]
```diff
@@ -118,6 +118,92 @@ struct NodeQuantizationInfo {
...
+    val_ = floatingToFixedPoint(floatVal, 32 - minBitsIntegerPart(floatVal));
+    intBits_ = minBitsIntegerPart(floatVal);
+    fracBits_ = 32 - intBits_;
+  };
+
+  /// Construct a fixed-point representation of the floating-point value
+  /// \p floatVal using the given number of integer bits \p intBits.
+  FixedPointUInt32(float floatVal, unsigned intBits) {
+    assert(floatVal >= 0 && "Floating point value must be positive!");
+    assert(intBits > 0 && intBits < 32 &&
```
Comment (mciprian13) [inline include/glow/Quantization/Base/Base.h:150]: Revise this assert.

File: lib/LLVMIRCodeGen/libjit/libjit.cpp [hunk 5]
```diff
@@ -2923,6 +2923,68 @@ void libjit_softmax_f(const float *inW, float *outW, const dim_t *idim,
   } // N
 }
 
+// multiply two fixed point number
+uint64_t multiply_fix_point(uint32_t a, uint32_t b) {
```
Comment (mciprian13) [inline lib/LLVMIRCodeGen/libjit/libjit.cpp:2927]: Perform this in-place.

File: lib/LLVMIRCodeGen/libjit/libjit.cpp [hunk 6]
```diff
@@ -2923,6 +2923,68 @@ void libjit_softmax_f(const float *inW, float *outW, const dim_t *idim,
...
+  uint64_t res = (uint64_t)a * (uint64_t)b;
+  return res;
+}
+
+// divide 2 fixed point number, indicating the fixed point format of the result
+uint32_t div_fixed_point(uint32_t x, uint32_t y,
```
Comment (mciprian13) [inline lib/LLVMIRCodeGen/libjit/libjit.cpp:2933]: Perform this in-place.

File: lib/LLVMIRCodeGen/libjit/libjit.cpp [hunk 7]
```diff
@@ -2923,6 +2923,68 @@ void libjit_softmax_f(const float *inW, float *outW, const dim_t *idim,
...
+
+  for (int j = 0; j < dims[0]; j++) {
+    uint32_t sum = 0;
+    int8_t max = std::numeric_limits<int8_t>::min();
+    uint32_t division;
+    int point, size;
+
+    // find max value
+    for (uint32_t i = 0; i < dims[1]; i++) {
+      max = MAX(max, inW[libjit_getXY(dims, j, i)]);
```
Comment (mciprian13) [inline lib/LLVMIRCodeGen/libjit/libjit.cpp:2951]: Try to use linear memory access and not the ```libjit_getXY``` primitive.

File: lib/LLVMIRCodeGen/LLVMIRGen.cpp [hunk 8]
```diff
@@ -2878,7 +2878,41 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
     auto *srcDims = emitValueDims(builder, src);
 
     auto *F = getFunction("softmax", dest->getElementType());
-    createCall(builder, F, {srcPtr, destPtr, srcDims, destDims});
+
+    if (src->getType()->isQuantizedType()) {
+      std::vector<int32_t> lut;
+
+      for (int32_t i = 0; i < 256; i++) {
+        lut.push_back((new FixedPointUInt32(
```
Comment (mciprian13) [inline lib/LLVMIRCodeGen/LLVMIRGen.cpp:2886]: Memory leak!

File: lib/LLVMIRCodeGen/LLVMIRGen.cpp [hunk 9]
```diff
@@ -2878,7 +2878,41 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
...
+      float size = static_cast<float>(src->getType()->dims()[1]);
+
+      auto *sumIntegerPart = emitConstI32(builder, ceil(log2(size)));
+
+      if (ceil(log2(size)) == floor(log2(size))) {
+        sumIntegerPart = emitConstI32(builder, ceil(log2(size)) + 1);
+      }
+
+      FixedPointUInt32 *invScaleFixedPoint =
+          new FixedPointUInt32(1.f / dest->getType()->getScale());
```
Comment (mciprian13) [inline lib/LLVMIRCodeGen/LLVMIRGen.cpp:2903]: Memory leak!

File: include/glow/Quantization/Base/Base.h (hunk 3 above)
Comment (EpureRares) [reply include/glow/Quantization/Base/Base.h]: Yes, I modified this part.

File: include/glow/Quantization/Base/Base.h (hunk 2 above)
Comment (EpureRares) [reply include/glow/Quantization/Base/Base.h]: To minimize the number of bits of the integer part of the fixed point representation, I modify the condition from the while block and the function returns integerPart

**Review by yinghai — APPROVED** [review]

**Pull request closed without merging**